Initial AD GenericMaterialProperty problem

vanwdani opened the pull request:
Refs #15213

Follow on PR for new AD template which causes qp index errors.
@bwspenc @lindsayad Do you have any insight?

vanwdani committed:
Initial AD GenericMaterialProperty problem
Refs #15213

lindsayad commented:
Can you summarize what the issue is that you're seeing?

bwspenc commented:
@lindsayad When we try to consolidate these classes (which grab components of tensors or do other reductions on them to provide a single Real value from a tensor) into a single templated class that uses `getGenericMaterialProperty()`, it seems like we're grabbing values from the wrong quadrature points. If we dump out the .e file to text (I think using exotxt or ncdump) and diff the new ones with ones generated from the gold files, you can see that we have all the right numbers, they're just reported for different elements.

lindsayad committed:
Don't bypass Material::get(AD)MaterialProperty

Before this commit `getGenericMaterialProperty` was skipping
straight through the `Material` class and going straight to the
`get(AD)MaterialProperty` methods in the `MaterialPropertyInterface`
class. This meant that for one thing `_requested_props` wasn't getting
set in the `Material` class, which meant that dependency resolution
was incorrect for any `Material` that employed
`getGenericlMaterialProperty`. With this change, dependency resolution
is correctly restored

lindsayad commented:
Ok @vanwdani @bwspenc 71e978f should hopefully fix things...

vanwdani committed:
Fix RankTwoScalarTools markdown
Refs #15214

vanwdani commented:
@lindsayad thanks, that worked great.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/519959/) on 4ca8093 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15214/site)

This comment will be updated on new commits.

bwspenc commented:
@lindsayad Thanks a lot for sorting that out!

@vanwdani Now that this issue is resolved, you should just make those changes we discussed yesterday (to use enumerations for the invariant type) on top of what you have here, rather than fixing it in the version that has two separate classes for AD and non-AD. I guess you could either put it in another commit in this PR or do a separate PR after this is merged, whichever you prefer.

vanwdani committed:
Use enum in RankTwo Material classes
Refs #15214

vanwdani commented:
@bwspenc I pulled this down last night and made the changes we discussed.  I'm pushing it up again, but I think I have a few more revisions I might want to make.

lindsayad commented:
@bwspenc I am quite confident that those aux variable failures in BISON triggered by the AD 2.0 change was due to this bug.

vanwdani committed:
Add additional strain/stress available to RankTwoTensorTools
Refs #15214

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/520561/) on c27b65a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15214/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15214/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 0b2836ad79fc9788b385406e68e10973a130be9c`

bwspenc commented:
@lindsayad I'm having trouble remembering what the AuxVariable problems were in BISON. We must have resolved them somehow. Maybe we should talk about that a little before we merge this.

bwspenc commented:
So if we have models with and without the `AD` prefix sharing the same implementation like this, do we end up not having a documentation page for the `AD` version? That's what this looks like. That's not completely a bad thing -- we don't want duplicate documentation pages floating around for no reason, but there should be some indication in the documentation system that the `AD` model exists and that the documentation for it is in the non-AD version.

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C:
```diff
@@ -18,38 +18,71 @@ const std::map<std::string, std::string>
-        {"max_principal", {"MaxPrincipal", {"stress"}}},
-        {"mid_principal", {"MidPrincipal", {"stress"}}},
-        {"min_principal", {"MinPrincipal", {"stress"}}}};
+        {"vonmises",
```
Whoa, hold on here!

Not every one of the invariants makes sense for every possible tensor, which is why we had a limited set defined originally. Some of these can be expanded to allow for other quantities to be computed, but not all of them, and we need to do this in a thoughtful way.  For now, please revert all of them back to the original set. 

Right off the top of my head, these are the obvious ones that could be expanded:
l2norm: It's OK for all of them
max_principal, mid_principal, min_principal: These should also be applicable to the strain
If we're going to add any, we need to test to make sure we've got them right, and add regression tests. For now, let's just leave them alone, and come back later and expand it.

vanwdani replied:
I set all these conditions to quickly convert over input files in BISON.  Files like RIA_CABRI_REP_Na4 have auxvariable/auxkernel pairs that follow the same convention as the already established sets.  (i.e. axial_creep_strain, hoop_elastic_strain, etc. )  I planned on converting over theses pairs, then removing the unnecessary tensors from the MOOSE PR.  I thought it would be faster to remove the tensors, than to collect all the necessary conditions, rebuild MOOSE, rebuild BISON, and then change the input files.

bwspenc replied:
I think those ones you mentioned make sense. Go ahead and submit another PR that adds the specific cases that are being used. I just don't think we can quite go and wholesale add every combination of options.

bwspenc commented on modules/tensor_mechanics/src/actions/TensorMechanicsActionBase.C:
```diff
@@ -18,38 +18,71 @@ const std::map<std::string, std::string>
 const std::map<std::string, std::pair<std::string, std::vector<std::string>>>
     TensorMechanicsActionBase::_rank_two_directional_component_table = {
-        {"directional", {"Direction", {"stress", "strain"}}}};
+        {"directional",
```
Same deal with all of these. Let's revert them back for now, and potentially add some as we add tests.

bwspenc commented on modules/tensor_mechanics/include/utils/RankTwoScalarTools.h:
```diff
@@ -27,6 +27,31 @@ MooseEnum invariantOptions();
+  MinPrincipal
+};
+
+enum class CYLINDRICAL_TYPE
```
This is getting picky, but I think it would be a little more clear if this were renamed to `CYLINDRICAL_COMPONENT`

bwspenc reviewed (changes requested):
This looks really good! Just a couple of little requested changes.

Once you address those minor things, I still have two questions from this that I'd like to discuss with the group:
1) Is there anything that these changes (Alex's stuff, specifically) would break in BISON, based on @lindsayad's comment?
2) How do we handle documentation for the AD and non-AD materials?

vanwdani commented:
@bwspenc I followed the no-AD convention of ComputeIsotropicElasticityTensor.md. This is a class that uses <is_ad> and the documention didn't mention AD.  I'm not sure if that was right or wrong

lindsayad commented:
> @lindsayad I'm having trouble remembering what the AuxVariable problems were in BISON. We must have resolved them somehow. Maybe we should talk about that a little before we merge this.

We saw exodiffs in a few tests so so we loosened the tolerances

bwspenc commented:
@lindsayad Well, if slightly loosening tolerances got them passing, then this shouldn't adversely affect them -- they'll just match the original solutions better. This is clearly an important fix.

bwspenc commented on modules/tensor_mechanics/doc/content/source/materials/ADRankTwoInvariant.md:
```diff
@@ -1,20 +1,20 @@
 ## Description
 
-This is an ADMaterial model used to extract an invariant of a rank-2 tensor in a
+This is a ADMaterial model used to extract an invariant of a rank-2 tensor in a
```
Umm, I'm pretty sure you use 'an' before a word that starts with an vowel.

I'll not hold up this PR for this, though.  Let's keep the AD and non-AD pages here for the time being, assuming the documentation builds fine, but we need to figure out a better long-term solution.

vanwdani replied:
Ha!  You are right.  I just added AD into the duplicate non-AD version.  It would make more sense to have the one md file that describes it's ability to do both things.

bwspenc commented:
The AD doc pages do show up and appear to be linked with AD classes, so until we come up with something better, I guess the thing to do is have doc pages for both classes.
